File: src-tauri/src/contracts.rs
use serde::{Deserialize, Serialize};

// ---------------------------------------------------------------------------
// NDJSON wire envelope (flat shape, per ndjson_examples.md)
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SidecarRequest {
    #[serde(rename = "id")]
    pub request_id: String,
    #[serde(rename = "type")]
    pub envelope_type: String,
    pub command: String,
    pub payload: serde_json::Value,
    #[serde(rename = "timestamp_ms")]
    pub timestamp_ms: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SidecarResponse {
    #[serde(rename = "request_id")]
    pub request_id: String,
    pub ok: bool,
    pub result: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SidecarErrorEnvelope {
    pub code: String,
    pub message: String,
    pub details: Option<serde_json::Value>,
}

// ---------------------------------------------------------------------------
// Sequence / annotation model
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SequenceState {
    pub id: String,
    pub name: String,
    pub sequence: String,
    pub topology: Topology,
    pub annotations: Vec<Annotation>,
    pub length_bp: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum Topology {
    Circular,
    Linear,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Annotation {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    pub annotation_type: String,
    pub start: usize,
    pub end: usize,
    pub strand: Strand,
    pub color: String,
    pub notes: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum Strand {
    Positive,
    Negative,
    Both,
}

// ---------------------------------------------------------------------------
// Parts & commands
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Part {
    pub id: String,
    pub name: String,
    pub role: String,
    pub sequence: String,
    pub annotations: Vec<Annotation>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "op", rename_all = "snake_case")]
pub enum Command {
    Insert { target_id: String, at: usize, inserted: Vec<Part> },
    Delete { target_id: String, from: usize, to: usize },
    Replace { target_id: String, from: usize, to: usize, replacement: Vec<Part> },
    AnnotationMove { annotation_id: String, new_start: usize, new_end: usize },
}

// ---------------------------------------------------------------------------
// Persistence / DB types
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DbSequence {
    pub id: String,
    pub name: String,
    pub sequence: String,
    pub topology: String,
    pub length_bp: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DbAnnotation {
    pub id: String,
    pub sequence_id: String,
    pub name: String,
    pub annotation_type: String,
    pub start: usize,
    pub end: usize,
    pub strand: String,
    pub color: String,
    pub notes: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DbPart {
    pub id: String,
    pub annotation_id: Option<String>,
    pub role: String,
    pub sequence: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DbAssemblyHistoryRow {
    pub id: String,
    pub sequence_id: String,
    pub command_json: String,
    pub created_at: String,
}

// ---------------------------------------------------------------------------
// Feature / tool contracts
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DigestRequest {
    pub id: String,
    pub sequence: String,
    pub enzymes: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DigestResponse {
    pub request_id: String,
    pub cuts: Vec<DigestCut>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DigestCut {
    pub enzyme: String,
    pub site: String,
    pub position: usize,
    pub fragment_length: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AutoAnnotateRequest {
    pub id: String,
    pub sequence: String,
    pub hints: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AutoAnnotateResponse {
    pub request_id: String,
    pub annotations: Vec<Annotation>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParseRequest {
    pub id: String,
    pub content: String,
    pub format: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParseResponse {
    pub request_id: String,
    pub sequence: SequenceState,
}

File: src/src/contracts.ts
// ===========================================================================
// Shared contract / spec types for NAipE (Sprint 1)
// React + Zustand consumer file — no app-level imports.
// ===========================================================================

// ---------------------------------------------------------------------------
// Error model
// ---------------------------------------------------------------------------

export interface AppError {
  error_code: ErrorCode;
  layer: 'Sidecar' | 'Core' | 'UI';
  message_dev: string;
  message_user: string;
  recoverable: boolean;
  context: Record<string, unknown>;
}

export interface SidecarResponse {
  request_id: string;
  ok: boolean;
  result: unknown;
}

export interface SidecarErrorEnvelope {
  code: string;
  message: string;
  details?: Record<string, unknown>;
}

export const enum ErrorCode {
  SIDECAR_TIMEOUT = 'SIDECAR_TIMEOUT',
  SIDECAR_CRASH = 'SIDECAR_CRASH',
  PARSE_INVALID_APE = 'PARSE_INVALID_APE',
  PARSE_INVALID_GENBANK = 'PARSE_INVALID_GENBANK',
  DB_WRITE_FAILED = 'DB_WRITE_FAILED',
  SEED_MISSING = 'SEED_MISSING',
  SEQUENCE_TOO_LARGE = 'SEQUENCE_TOO_LARGE',
}

// ---------------------------------------------------------------------------
// Sequence / annotation model
// ---------------------------------------------------------------------------

export type Topology = 'circular' | 'linear';
export type Strand = '+' | '-' | 'both';

export interface Annotation {
  id: string;
  name: string;
  type: string;
  start: number;
  end: number;
  strand: Strand;
  color: string;
  notes: string;
}

export interface SequenceState {
  id: string;
  name: string;
  sequence: string;
  topology: Topology;
  annotations: Annotation[];
  length_bp: number;
}

// ---------------------------------------------------------------------------
// Parts & commands
// ---------------------------------------------------------------------------

export interface Part {
  id: string;
  name: string;
  role: string;
  sequence: string;
  annotations: Annotation[];
}

export type Command =
  | { op: 'insert'; target_id: string; at: number; inserted: Part[] }
  | { op: 'delete'; target_id: string; from: number; to: number }
  | { op: 'replace'; target_id: string; from: number; to: number; replacement: Part[] }
  | { op: 'annotation_move'; annotation_id: string; new_start: number; new_end: number };

// ---------------------------------------------------------------------------
// Persistence / DB types
// ---------------------------------------------------------------------------

export interface DbSequence {
  id: string;
  name: string;
  sequence: string;
  topology: string;
  length_bp: number;
}

export interface DbAnnotation {
  id: string;
  sequence_id: string;
  name: string;
  type: string;
  start: number;
  end: number;
  strand: string;
  color: string;
  notes: string;
}

export interface DbPart {
  id: string;
  annotation_id: string | null;
  role: string;
  sequence: string;
}

export interface DbAssemblyHistoryRow {
  id: string;
  sequence_id: string;
  command_json: string;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Feature / tool contracts
// ---------------------------------------------------------------------------

export interface DigestRequest {
  id: string;
  sequence: string;
  enzymes: string[];
}

export interface DigestCut {
  enzyme: string;
  site: string;
  position: number;
  fragment_length: number;
}

export interface DigestResponse {
  request_id: string;
  cuts: DigestCut[];
}

export interface AutoAnnotateRequest {
  id: string;
  sequence: string;
  hints: string[];
}

export interface AutoAnnotateResponse {
  request_id: string;
  annotations: Annotation[];
}

export interface ParseRequest {
  id: string;
  content: string;
  format: string;
}

export interface ParseResponse {
  request_id: string;
  sequence: SequenceState;
}
